core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java

```diff
@@ -109,6 +113,67 @@ public DefaultRequestLog(RequestContext ctx) {
         this.ctx = requireNonNull(ctx, "ctx");
     }
 
+    @Override
+    public void addChild(RequestLog child, boolean lastChild) {
+        checkState(!hasLastChild, "last child is already added");
+        requireNonNull(child, "child");
+        if (children == null) {
+            // first child's all request-side logging events are propagated immediately to the parent
+            children = new ArrayList<>();
+            propagateRequestSideLog(child);
+        }
+        children.add(child);
+        if (lastChild) {
+            setRecentAsLastChild();
+        }
+    }
+
+    private void propagateRequestSideLog(RequestLog child) {
+        child.addListener(log -> startRequest(log.channel(), log.sessionProtocol(), log.host()), REQUEST_START);
+        child.addListener(log -> serializationFormat(log.serializationFormat()), SCHEME);
+        child.addListener(log -> requestHeaders(log.requestHeaders()), REQUEST_HEADERS);
+        child.addListener(log -> requestContent(log.requestContent(), log.rawRequestContent()),
+                          REQUEST_CONTENT);
+        child.addListener(log -> {
+            if (log.requestCause() != null) {
+                endRequest(log.requestCause());
+            } else {
+                endRequest();
+            }
+        }, REQUEST_END);
+    }
+
+    @Override
+    public void setRecentAsLastChild() {
+        checkState(!hasLastChild, "last child is already added");
+        checkState(!children.isEmpty(), "at least one child should be already added");
+        hasLastChild = true;
+        final RequestLog lastChild = children.get(children.size() - 1);
+        propagateResponseSideLogWhenComplete(lastChild);
+    }
+
+    private void propagateResponseSideLogWhenComplete(RequestLog lastChild) {
+        lastChild.addListener(log -> {
+            if (log instanceof DefaultRequestLog) {
+                startResponse0(true, ((DefaultRequestLog) log).responseStartTimeNanos,
+                               log.responseStartTimeMillis());
+            } else {
+                startResponse();
+            }
+            if (log.responseHeaders() != HttpHeaders.EMPTY_HEADERS) {
+                responseHeaders(log.responseHeaders());
+            }
+            if (log.responseContent() != null) {
+                responseContent(log.responseContent(), log.rawResponseContent());
+            }
+            if (log.responseCause() != null) {
+                endResponse(log.responseCause());
+            } else {
+                endResponse();
+            }
+        }, COMPLETE);
```

Comment:
Doesn't need to use `COMPLETE` but just `RESPONSE_END`.

Reply:
All of the codes are fixed. Thank you~!